node/src/connection.rs

```diff
@@ -0,0 +1,62 @@
+use std::collections::HashMap;
+use std::time::{Duration, Instant};
+
+use cait_sith::protocol::Participant;
+use tokio::sync::RwLock;
+
+use crate::protocol::contract::primitives::Participants;
+use crate::web::StateView;
+
+const DEFAULT_TIMEOUT: Duration = Duration::from_secs(5);
+
+// TODO: this is a basic connection pool and does not do most of the work yet. This is
+//       mostly here just to facilitate offline node handling for now.
+// TODO/NOTE: we can use libp2p to facilitate most the of low level TCP connection work.
+pub struct Pool {
+    connections: RwLock<HashMap<Participant, String>>,
+    http: reqwest::Client,
+    current_active: RwLock<Option<(Vec<Participant>, Instant)>>,
+}
+
+impl Pool {
+    pub fn spawn() -> Self {
+        Self {
+            connections: Default::default(),
+            http: reqwest::Client::default(),
+            current_active: RwLock::new(None),
+        }
+    }
+
+    pub async fn set_participants(&self, participants: &Participants) {
+        let mut connections = self.connections.write().await;
+        *connections = participants
+            .iter()
+            .map(|(p, info)| (*p, info.url.clone()))
+            .collect();
+    }
+
+    pub async fn ping(&self) -> Vec<Participant> {
+        if let Some((ref active, timestamp)) = *self.current_active.read().await {
+            if timestamp.elapsed() < DEFAULT_TIMEOUT {
+                return active.clone();
+            }
+        }
+
+        let connections = self.connections.read().await;
+        let mut participants = vec![];
+        for (participant, url) in connections.iter() {
+            let Ok(resp) = self.http.get(format!("{url}/state")).send().await else {
```

Comment:
Do we need to set a timeout here? Would it hang forever?

Reply:
The reqwest client should have a default 30 second timeout, so it should be fine